src/std_extensions/quantum.rs

```diff
@@ -1,19 +1,190 @@
+    TypeArg::BoundedNat {
+        n: log_denom as u64,
+    }
+}
+
+/// An angle
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct ConstAngle {
+    log_denom: u8,
+    value: u64,
+}
+
+impl ConstAngle {
+    /// Create a new [`ConstAngle`] from a log-denominator and a numerator
+    pub fn new(log_denom: u8, value: u64) -> Result<Self, ConstTypeError> {
+        if !is_valid_log_denom(log_denom) {
+            return Err(ConstTypeError::CustomCheckFail(
+                crate::types::CustomCheckFailure::Message(
+                    "Invalid angle log-denominator.".to_owned(),
+                ),
+            ));
+        }
+        if value >= (1u64 << log_denom) {
+            return Err(ConstTypeError::CustomCheckFail(
+                crate::types::CustomCheckFailure::Message(
+                    "Invalid unsigned integer value.".to_owned(),
+                ),
+            ));
+        }
+        Ok(Self { log_denom, value })
+    }
+
+    /// Create a new [`ConstAngle`] from a log-denominator and a floating-point value in radians,
+    /// rounding to the nearest corresponding value. (Ties round away from zero.)
+    pub fn from_radians(log_denom: u8, theta: f64) -> Result<Self, ConstTypeError> {
+        if !is_valid_log_denom(log_denom) {
+            return Err(ConstTypeError::CustomCheckFail(
+                crate::types::CustomCheckFailure::Message(
+                    "Invalid angle log-denominator.".to_owned(),
+                ),
+            ));
+        }
+        let a = (((1u64 << log_denom) as f64) * theta / TAU).round() as i64;
+        Ok(Self {
+            log_denom,
+            value: a.rem_euclid(1i64 << log_denom) as u64,
+        })
+    }
+
+    /// Returns the value of the constant
+    pub fn value(&self) -> u64 {
+        self.value
+    }
+
+    /// Returns the log-denominator of the constant
+    pub fn log_denom(&self) -> u8 {
+        self.log_denom
+    }
+}
+
+#[typetag::serde]
+impl CustomConst for ConstAngle {
+    fn name(&self) -> SmolStr {
+        format!("a(2π*{}/2^{})", self.value, self.log_denom).into()
+    }
+    fn check_custom_type(&self, typ: &CustomType) -> Result<(), CustomCheckFailure> {
+        if typ.clone() == angle_custom_type(type_arg(self.log_denom)) {
+            Ok(())
+        } else {
+            Err(CustomCheckFailure::Message(
+                "Angle constant type mismatch.".into(),
+            ))
+        }
+    }
+    fn equal_consts(&self, other: &dyn CustomConst) -> bool {
+        crate::values::downcast_equal_consts(self, other)
+    }
+}
+
+fn atrunc_sig(arg_values: &[TypeArg]) -> Result<FunctionType, SignatureError> {
+    let [arg0, arg1] = collect_array(arg_values);
+    let m: u8 = get_log_denom(arg0)?;
+    let n: u8 = get_log_denom(arg1)?;
+    if m < n {
+        return Err(SignatureError::InvalidTypeArgs);
+    }
+    Ok(FunctionType::new(
+        vec![angle_type(arg0.clone())],
```

Comment:
Nit - do you need to clone these? (not sure why `clippy` doesn't complain?!)

Reply:
Without `clone()` I get an error:

```
error[E0308]: mismatched types
   --> src/std_extensions/quantum.rs:153:25
    |
153 |         vec![angle_type(arg0)],
    |              ---------- ^^^^ expected `TypeArg`, found `&TypeArg`
    |              |
    |              arguments to this function are incorrect
    |
note: function defined here
   --> src/std_extensions/quantum.rs:35:15
    |
35  | pub(super) fn angle_type(log_denom_arg: TypeArg) -> Type {
    |               ^^^^^^^^^^ ----------------------
help: consider using clone here
    |
153 |         vec![angle_type(arg0.clone())],
```

and the suggested fix is to use `clone()`.

Reply:
Ok - think I was forgetting how slices work, sorry!